### ISSUE TITLE
Please let me know when the backend is finished

### ISSUE THREAD

**Issue by RomanPolek** [RomanPolek/wad2_ae1#15]
Hey backend team,
Please let me know as soon as the backend is finished. We have to then connect the backend frontend together and fix bugs together. Ill not be working on backend anymore as I see that you started to work on it. There's very little time left, that's why while you are working on it and frontend is waiting, Im gonna fork and work on a backup - in case the main branch doesnt work near the deadline, we can use the working finished backup branch.

We should definitely have a meeting and discuss the progress. Just let me know when u can today @mrkajetanp @2397741k

**Comment by mrkajetanp** [comment]
@RomanPolek Yesterday I added an additional role, 'None' as the default role so that superusers aren't lumped in with Students anymore. It seems to have fixed the issues around it and doesn't affect user registration as the form only has the Student/Teacher choice anyway. As far as I can tell it didn't break anything but if you have some code somewhere that relies on Student being the default role, let me know / take a look at it!

Also, is there anything specific you need me to fix/implement?

**Comment by RomanPolek** [comment]
Thanks, It did not break anything.
Could you please look at a way to implement edit / remove exam? I imagine in a teachers view of /exams/ page, each exam would have 3 buttons - view edit and delete. Then I would just reuse the template for add exam, but fill it with existing exam data and change other bits and it should work. So could you please have a look at this? Dont worry about the button positioning / style too much, just so the functionality is there. Ill then make the buttons neater in the frontend. Thanks. While doing this please try to use existing functions. get_courses() will get all available courses for the person - performs checks. get_exams(courses) will get all available exams. error displays errors etc.

Here I imagine the buttons 3 would be only for teacher. Student will have 1 button - Start as there is alrdy:
![image](https://user-images.githubusercontent.com/42897056/113594696-f1d79980-962f-11eb-928a-31cd869a1d7f.png)
![image](https://user-images.githubusercontent.com/42897056/113594857-221f3800-9630-11eb-9ef4-c2105cf89605.png)

**Comment by mrkajetanp** [comment]
Some other things we need to fix / implement that I found along the way:
- teachers & students can change their role at will, that should be fixed
- should people be able to create teacher accounts without any sort of verification? should that somehow be addressed?
- submitting an exam doesn't add a submission in the db, should be implemented (I'm pretty sure you know that but just noting)
- the server is throwing warnings about timezones which related to timezone support being turned on but not actually used in the code, should we just turn timezone support off and not bother? I don't think it really matters tbh
- views.py contains a view called exams and show_exams with the latter seemingly not used at all, do we need it for anything?

**Comment by RomanPolek** [comment]
1) Ill fix this
2) Somebody please look into this, but at this point i dont think we have time to implement this
3) Im working on it rn
4) I tried to fix it, idk why it still shows this warning. This is low priority fix - we dont use timezones anyways
5) Ill fix this
Thanks

**Comment by RomanPolek** [comment]
I alrdy fixed 1., 5.
Ill continue working on submitting the exam / calculating statistics / displaying mistakes to students etc..

**Comment by mrkajetanp** [comment]
When adding a exam through the front end, the form doesn't let you add a question with any of the 5 answers missing. That should definitely be fixed.

Teachers can't view the exams that are not currently open, I'm pretty sure it's due to the check in the render call at the beginning of exams view.

**Comment by RomanPolek** [comment]
1) yep for now not. This is medium priority. I wanna finish firstly so that user needs to put every answer cuz its easier then if we have time ill change it

2) Its due to the get_exams which checks for deadline. U can add a parameter to the function which overrides this check or sth

**Comment by mrkajetanp** [comment]
@RomanPolek If you look at exams.html, for some reason changing anything about the <a> tags there completely breaks the formatting of specific exams showing up in exams view. To make specific buttons do different things I had to wrap just them in <a>s but that had really weird effects on positioning of things. I market the relevant parts with "BUG" in html comments, could you take a look?

**Comment by RomanPolek** [comment]
ye i know about this. The exam containers are kinda buggy. ill polish them as soon as u finish. Dont worry about the frontend too much because ill be changing it slightly anyways

**Comment by RomanPolek** [comment]
OK i made a big commit. I added the submissions etc. all around submitting exams. It shows u which answers wrong, score, preview of past exams...

**Comment by mrkajetanp** [comment]
That's great! I've got editing exams mostly worked out, I'll commit it soonish after I eat dinner

**Comment by RomanPolek** [comment]
perfect. and ill also soon commit the statistics thing for class

**Comment by RomanPolek** [comment]
@mrkajetanp Anna wanted to move the meeting to 10:00. She needs more time to finish. So the meeting will be at 10 then. Therefore we also have time to polish everything

**Comment by mrkajetanp** [comment]
@RomanPolek All right, both removing and editing exams now works, what do you need me to do next?

**Comment by RomanPolek** [comment]
Great job! Now in the view exams I calculate the percentage etc.. create new submission
I think it would be better if we calculated this directly in the model and when exam is updated it will automatically update all the submissions.
Imagine a teacher publishes some exam, but then finds out one question is bad so she edits it out - after some ppl alrdy finished. As soon as she removes the question from the exam all the submissions linked to the exam will update. I would put some method directly inside the submission model e.g. recalculate, which would calculate the scores max scores, percentages.. again. I think the Answer class will get deleted by CASCADE when the question is deleted

**Comment by RomanPolek** [comment]
Also we need to handle what if teacher deletes the exam - then all the submission will prolly need to be deleted too

**Comment by RomanPolek** [comment]
this might help u. Its a receiver that listens for updates of User. So we can do the same with Exam and Submissions
![image](https://user-images.githubusercontent.com/42897056/113619415-da5cd880-9650-11eb-8529-bc07aeefd323.png)

**Comment by mrkajetanp** [comment]
Seems like the main thing we're missing here besides that is letting teachers see their students' exam results when they click 'view' on an exam, should I just do that first?

**Comment by RomanPolek** [comment]
yep sure thats a great idea. Actually I can do this, because I already have everything nicely setup for this. It will be easy for me

**Comment by RomanPolek** [comment]
So u can start work on the first thing im gonna now do the thing u said

**Comment by mrkajetanp** [comment]
All right all right

**Comment by mrkajetanp** [comment]
> Also we need to handle what if teacher deletes the exam - then all the submission will prolly need to be deleted too

Submissions already delete themselves if the exam gets removed due to cascading, I just tested it so we've got that covered

**Comment by RomanPolek** [comment]
ye i mean mainly calculate percentage, score, max_score vars in Submission
Cuz if u remove question and Answer, percentage, score max_score changes

**Comment by RomanPolek** [comment]
btw great job on the Edit exam button. Works flawlessly and looks amazing

**Comment by RomanPolek** [comment]
![image](https://user-images.githubusercontent.com/42897056/113620597-72a78d00-9652-11eb-8b68-22ec136d2a77.png)

**Comment by mrkajetanp** [comment]
Thanks! I had to write some slightly weird JS but if it works it works I suppose

**Comment by RomanPolek** [comment]
i updated backend so get_exams() function returns all exams for teachers (also before and past deadline)

**Comment by RomanPolek** [comment]
and finished the viewing of students exams by teacher

**Comment by mrkajetanp** [comment]
That's great! I'll have updating the submissions working pretty soon, already got most of it

**Comment by RomanPolek** [comment]
im working on the statistics container now. should be finished before meeting

**Comment by RomanPolek** [comment]
teachers will have modified statistics container where they will also see each person in this course

**Comment by RomanPolek** [comment]
it would be wise to have a remove button next to each person. so teachers can remove ppl from course

**Comment by mrkajetanp** [comment]
The submissions now update after their exam gets edited.
I've noticed a backend issue that doesn't affect the usability in any way, namely, because questions are connected to exams with a many-to-many field they don't get deleted along with the exam leading to old questions being left in the db for no reason. Replacing it with a foreign key relationship should fix it, should I try doing that?

**Comment by RomanPolek** [comment]
https://stackoverflow.com/questions/3937194/django-cascade-deletion-in-manytomanyrelation

**Comment by RomanPolek** [comment]
![image](https://user-images.githubusercontent.com/42897056/113625955-7a1e6480-9659-11eb-8d5b-7f07da7367df.png)

**Comment by RomanPolek** [comment]
i personally think M2M is ok. We can do a receiver that will delete like this. Also foreign keys would work but i think M2M fits a bit better if we can solve that issue with this solution

**Comment by mrkajetanp** [comment]
Oh that's even better, thanks! I'll implement it then

**Comment by mrkajetanp** [comment]
@RomanPolek I'm not sure if it's just my machine or not but in add_course the Description field seems to be a bit too wide and sticking outside the containing rectangle, could you take a look at it?

**Comment by mrkajetanp** [comment]
@RomanPolek If you try doing an exam as a student the page still seems to highlight the correct answer which is obviously a problem, would you be okay fixing that?

**Comment by RomanPolek** [comment]
solved

**Comment by RomanPolek** [comment]
im just testing it and pushed the changes rn :)

**Comment by RomanPolek** [comment]
also I tested the Submission updated and works perfectly. Thanks great job

**Comment by mrkajetanp** [comment]
I implemented showing Students their past submitted exams in Exam view!
Essentially I just changed the get_exams function to perform two queries (submitted & upcoming exams) and join them into one, resulting in past exams only showing up if they have a submission from this student and upcoming showing up same as before. No change to what Teachers get.

I also removed an unnecessary check since you were filtering exams by dates in exams view having already filtered them in get_exams(), if an exam got that far it should mean it's all right!

**Comment by 2397741k** [comment]
That’s great!
